### PR TITLE
WIP: Event recorder

### PR DESCRIPTION
**Pull request by jetuk**
Created PR for discussion.

Currently does the following,
- Adds EventRecorder and Event object.
- Basic test for capturing the correct events on a scenario basis.
- Creates a pywr.parameters._thresholds module for threshold parameters.
- Moved RecorderThresholdParameter to new module.
- Added NodeThresholdParameter and StorageThresholdParameter.

Todo 
- Track other things during an event.

Also #323 would break some things in here.

### CODE REVIEW

File: pywr/recorders/events.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+     See also
+     --------
+     `pywr.parameters._thresholds`
+
+
+     """
+    def __init__(self, model, threshold, *args, **kwargs):
+        super(EventRecorder, self).__init__(model, *args, **kwargs)
+        self.threshold = threshold
```
Comment (jetuk) [inline pywr/recorders/events.py]: Add a TODO comment regarding setting parent relationship with `self.threshold`

File: pywr/parameters/_thresholds.pyx [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+        return self.recorder.data[timestep._index - 1, scenario_index._global_id]
+
+    cpdef int index(self, Timestep timestep, ScenarioIndex scenario_index) except? -1:
+        """Returns 1 if the predicate evalutes True, else 0"""
+        cdef int index = timestep.index
+        cdef int ind
+        if index == 0:
+            # on the first day the recorder doesn't have a value so we have no
+            # threshold to compare to
+            ind = 1
```
Comment (snorfalorpagus) [inline pywr/parameters/_thresholds.pyx]: This should be configurable. `self.initial_value`?

File: pywr/recorders/_recorders.pxd [hunk 3]
```diff
@@ -6,6 +6,7 @@ from pywr.parameters._parameters cimport Parameter, IndexParameter
 cdef class Recorder(Component):
     cdef bint _is_objective
     cdef public double epsilon
+    cdef public double ignore_nan
```
Comment (snorfalorpagus) [inline pywr/recorders/_recorders.pxd]: This is the wrong type. Should be `bint`.

File: pywr/parameters/_thresholds.pyx [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+    """
+
+    def __init__(self,  model, Recorder recorder, *args, **kwargs):
+        super(RecorderThresholdParameter, self).__init__(model, *args, **kwargs)
+        self.recorder = recorder
+        self.recorder.parents.add(self)
+
+    cpdef double _value_to_compare(self, Timestep timestep, ScenarioIndex scenario_index) except? -1:
+        # TODO Make this a more general API on Recorder
+        return self.recorder.data[timestep._index - 1, scenario_index._global_id]
```
Comment (snorfalorpagus) [inline pywr/parameters/_thresholds.pyx]: Should this use `self.recorder.values()`?

File: pywr/recorders/events.py (hunk 1 above)
Comment (snorfalorpagus) [inline pywr/recorders/events.py]: This has been done.

File: pywr/recorders/events.py [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+        # Optional different method for aggregating across self.recorders scenarios
+        agg_func = kwargs.pop('recorder_agg_func', kwargs.get('agg_func'))
+        self.recorder_agg_func = agg_func
+
+        super(EventDurationRecorder, self).__init__(model, **kwargs)
+        self.event_recorder = event_recorder
+        self.event_recorder.parents.add(self)
+
+    def setup(self):
+        self._values = np.zeros(len(self.model.scenarios.combinations))
```
Comment (snorfalorpagus) [inline pywr/recorders/events.py]: This can use `np.empty` as `setup` calls `reset`.

File: pywr/parameters/_thresholds.pyx (hunk 4 above)
Comment (snorfalorpagus) [inline pywr/parameters/_thresholds.pyx]: I'm currently using this recorder with "MeanFlowRecorder". I would need to implement `.values` on that, but it shouldn't be difficult.

Comment (jetuk) [inline pywr/parameters/_thresholds.pyx]: Well I am not sure about this. `Recorder.values` is currently intended to produce a single value for each scenario at the end of the model run. We've not used as a general API for getting a value from a `Recorder` during a run.

**Pull request merged**